derp/derp_client.go

```diff
@@ -98,46 +105,40 @@ func (c *Client) recvServerInfo() (*serverInfo, error) {
 }
 
 func (c *Client) sendClientKey() error {
-	var nonce [24]byte
+	var nonce [nonceLen]byte
 	if _, err := crand.Read(nonce[:]); err != nil {
 		return err
 	}
 	msg := []byte("{}") // no clientInfo for now
 	msgbox := box.Seal(nil, msg, &nonce, c.serverKey.B32(), c.privateKey.B32())
 
-	if _, err := c.bw.Write(c.publicKey[:]); err != nil {
-		return err
-	}
-	if _, err := c.bw.Write(nonce[:]); err != nil {
-		return err
-	}
-	if err := putUint32(c.bw, uint32(len(msgbox))); err != nil {
-		return err
-	}
-	if _, err := c.bw.Write(msgbox); err != nil {
-		return err
-	}
-	return c.bw.Flush()
+	buf := make([]byte, 0, nonceLen+keyLen+len(msgbox))
+	buf = append(buf, c.publicKey[:]...)
+	buf = append(buf, nonce[:]...)
+	buf = append(buf, msgbox...)
+	return writeFrame(c.bw, frameClientInfo, buf)
 }
 
-func (c *Client) Send(dstKey key.Public, msg []byte) (err error) {
+func (c *Client) Send(dstKey key.Public, msg []byte) error { return c.send(dstKey, msg) }
+
+func (c *Client) send(dstKey key.Public, msg []byte) (ret error) {
 	defer func() {
-		if err != nil {
-			err = fmt.Errorf("derp.Send: %v", err)
+		if ret != nil {
+			ret = fmt.Errorf("derp.Send: %v", ret)
 		}
 	}()
 
-	if err := typeSendPacket.Write(c.bw); err != nil {
-		return err
+	// Verify len(msg) + 4 fits in a uint32 with a sanity check.
+	// This isn't a real limit, but is much higher than a IP
+	// packet size.
+	if len(msg) > 10<<20 {
```

Comment:
This could be a lot smaller right? We move UDP packets, so the limit is 64kb. 1<<17?

Reply:
Done. Capped and documented at 64<<10. (Let's only use shifts of 10, 20, 30 like Go)